lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -4229,6 +4244,10 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
   }
 
   void visitStructExtractInst(StructExtractInst *sei) {
+    // There does not exist a corresponding cotangent field for original
+    // fields with `@noDerivative` attribute. Emit an error.
+    if (sei->getField()->getAttrs().hasAttribute<NoDerivativeAttr>())
```

Comment:
I don't think you need this anymore. I believe an assertion is safe.

Reply:
Assertion added in https://github.com/apple/swift/pull/22973/commits/37c6620662c267ae2cd8665f4d8d620fbec8192a.

Verified with sample programs that assertion is never triggered, because activity analysis does not propagate variedness for `struct_extract` instructions with `@noDerivative` fields.